clientd3d/gameuser.c

```diff
@@ -147,13 +149,29 @@ void GotObjectContents(ID object_id, list_type contents)
 	 GameMessagePrintf(GetString(hInst, IDS_EMPTY), LookupNameRsc(r->obj.name_res));
       return;
    }
+   //Separate contents into number items and other items
+   for (l = contents; l != NULL; l = l->next)
+   {
+      if (IsNumberObj(((object_node *)(l->data))->id))
+      {
+         stacklist = list_add_sorted_item(stacklist, (object_node *)(l->data), CompareObjectNameRsc);
+      }
+      else
+      {
+         nostacklist = list_add_sorted_item(nostacklist, (object_node *)(l->data), CompareObjectNameRsc);
+      }
+   }
+   //Concat the two lists into one
+   stacklist = list_append(stacklist, nostacklist);
```

Comment:
Rather than building up two lists and concatenating them, it will be shorter and easier to understand if you just do

```
   for (l = contents; l != NULL; l = l->next)
      sorted_list = list_add_sorted_item(sorted_list, (object_node *) (l->data), CompareObjectNameAndNumber);
```

where the new comparison proc works this way:

- Compute two variables (say, `number1` and `number2` to determine whether the items are number items.
- If number1 != number2, return a value (1 or -1) that puts the number item first.
- Otherwise, look up the two names and compare them, like you do in CompareObjectNameRsc.

Make sure that this new comparison proc has a comment describing its purpose (sort number items before other items, and within each group, sort alphabetically).

Reply:
Good idea and done.